models/user/userModel.js

```diff
@@ -0,0 +1,84 @@
+import mongoose, { Schema } from 'mongoose';
+import { addressSchema, orderItemSchema } from '../schemas/index.js';
+import { accountStatuses } from '../../constants/index.js';
+import {
+  emailPattern,
+  passwordPattern,
+  phoneNumberPattern,
+} from '../../helpers/index.js';
+import { roleSchema, getDefaultRoles } from './roleSchema.js';
+
+const ObjectId = Schema.Types.ObjectId;
+
+const userSchema = new Schema(
+  {
+    firstName: {
+      type: String,
+      required: [true, 'First name is required'],
+      minlength: [3, 'First name must be at least 3 characters long'],
+    },
+    lastName: {
+      type: String,
+      required: [true, 'Last name is required'],
+      minlength: [3, 'Last name must be at least 3 characters long'],
+    },
+    password: {
+      type: String,
+      required: [true, 'Password is required'],
+      validate: {
+        validator: (value) => passwordPattern.test(value),
+        message: 'Invalid password format',
+      },
+      minlength: [8, 'Password must be at least 8 characters long'],
+    },
+    email: {
+      type: String,
+      required: [true, 'Email is required'],
+      unique: true,
+      validate: {
+        validator: (value) => emailPattern.test(value),
+        message: 'Invalid email format',
+      },
+    },
+    avatar: String,
+    address: addressSchema,
+    phoneNumber: {
+      type: String,
+      validate: {
+        validator: (value) => phoneNumberPattern.test(value),
+        message: 'Invalid phone number format',
+      },
+    },
+    favoriteDishes: [{ type: ObjectId, ref: 'dish' }],
+    favoriteChefs: [{ type: ObjectId, ref: 'chef' }],
+    cart: [orderItemSchema],
+    roles: {
+      type: [roleSchema],
+      default: getDefaultRoles,
```

Comment:
so by default user will have all the roles available?
looks like potential security issue for me, you should use least privilege principle - grant minimal required amount of permissions for account to do its job

Reply:
By default a user will have only one role - USER. That's why I added `{id : null}` for other possible roles. And if he registers for another role, the id of that role is added to the `roles` array.

But I got your point, and will change it so that the default value is `roles: [ ]`.

Reply:
I made the default `roles: [{"user" : "userId"}]` in order to return only one object ` req.roleIds` from the auth middleware instead of two objects, as it was in the previous version of the middleware (`req.userId = `, `req.roles =` )

right now getDefaultRoles has the next view
`export const getDefaultRoles = function () {
  return [
    {
      name: roles.USER,
      id: this._id,
    },
  ];
};`